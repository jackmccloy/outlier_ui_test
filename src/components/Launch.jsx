import React, { Component } from 'react';

class Launch extends Component {

  render() {

    let launch = this.props.launch;

    return (
      <li>
        <h2> { launch.mission_name } </h2>
        <div> { launch.rocket.rocket_name } </div>
        <div > { launch.rocket.rocket_id } </div>
      </li>
    );
  }
}

export default Launch;