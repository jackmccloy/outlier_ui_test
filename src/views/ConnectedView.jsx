import React, { Component } from 'react';
import { connect } from "react-redux";

import Navigation from '../components/Navigation';

const menu = Navigation();

function MasterLayoutHOC(WrappedComponent, pageName) {
  class MasterLayoutImpl extends Component {
    render() {

      const layoutProps = {
        menu,
        pageName
      };

      return (
        <WrappedComponent {...this.props} />
      );
    }
  }

  const mapStateToProps = state => state;

  const mapDispatchToProps = dispatch => ({
    dispatch
  });

  return connect(
    mapStateToProps,
    mapDispatchToProps
  )(MasterLayoutImpl);
}

export default MasterLayoutHOC;
