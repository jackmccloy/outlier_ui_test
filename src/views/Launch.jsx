import React, { Component } from 'react';

import { Link } from 'react-router-dom';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getLaunchesThunk } from "../actions/launches";
import ImagesArea from '../components/ImagesArea';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const LaunchH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class LaunchView extends Component {
  componentDidMount() {
    const { getLaunches } = this.props;
    getLaunches();
  }

  render() {
    const {
      match,
      launchCollection,
    } = this.props;

    if (!launchCollection || launchCollection.fetching) {
      return (
        <Section>
          <LaunchH1> SpaceX </LaunchH1>
          <div> LOADING... </div>
        </Section>
      );
    }

    if (launchCollection.error) {
      return (
        <Section>
          <LaunchH1> SpaceX </LaunchH1>
          <div> { launchCollection.error } </div>
        </Section>
      );
    }

    const launchId = parseInt(match.params.flight_number, 10);
    const thisLaunch = launchCollection.launches.filter(launch => launch.flight_number === launchId);

    if (thisLaunch.length !== 1) {
      return (
        <Section>
          <LaunchH1> SpaceX </LaunchH1>
          <div> NO DATA </div>
        </Section>
      );
    }

    return (
      <Section>
        <LaunchH1> SpaceX { thisLaunch[0].mission_name }, {thisLaunch[0].launch_year}</LaunchH1>
        <h2> Rocket: <Link to={`/rockets/${thisLaunch[0].rocket.rocket_id}`}>
          { thisLaunch[0].rocket.rocket_name }</Link>
        </h2>
        <div> { thisLaunch[0].details } </div>
        <ImagesArea
          title={thisLaunch[0].mission_name}
          images={thisLaunch[0].links.flickr_images}
        />
      </Section>
    );
  }
}

LaunchView.propTypes = {
  launchCollection: PropTypes.shape({
    launches: PropTypes.arrayOf(PropTypes.shape({
      flight_number: PropTypes.string.isRequired,
      mission_name: PropTypes.string,
      launch_year: PropTypes.string,
      details: PropTypes.string,
      rocket: PropTypes.shape({
        rocket_id: PropTypes.string,
        rocket_name: PropTypes.string,
      }),
      links: PropTypes.shape({
        flickr_images: PropTypes.arrayOf(PropTypes.string),
      }),
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getLaunches: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      flight_number: PropTypes.string,
    }),
  }).isRequired,
}

const mapStateToProps = state => ({
  launchCollection: state.launchCollection,
});

const mapDispatchToProps = {
  getLaunches: getLaunchesThunk,
};

const ConnectedLaunchView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(LaunchView);

export default ConnectedLaunchView;
