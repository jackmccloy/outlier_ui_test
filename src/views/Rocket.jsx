import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getRocketsThunk } from "../actions/rockets";
import ImagesArea from '../components/ImagesArea';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const RocketH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class RocketView extends Component {
  componentDidMount() {
    const { getRockets } = this.props;
    getRockets();
  }

  render() {
    const {
      match,
      rocketCollection,
    } = this.props;

    if (!rocketCollection || rocketCollection.fetching) {
      return (
        <Section>
          <RocketH1> SpaceX </RocketH1>
          <div> LOADING... </div>
        </Section>
      );
    }

    if (rocketCollection.error) {
      return (
        <Section>
          <RocketH1> SpaceX </RocketH1>
          <div> { rocketCollection.error } </div>
        </Section>
      );
    }

    const rocketId = match.params.rocket_id;
    const thisRocket = rocketCollection.rockets.filter(rocket => rocket.rocket_id === rocketId);

    if (thisRocket.length !== 1) {
      return (
        <Section>
          <RocketH1> SpaceX </RocketH1>
          <div> NO DATA </div>
        </Section>
      );
    }

    return (
      <Section>
        <RocketH1> SpaceX { thisRocket[0].rocket_name } </RocketH1>
        <div> { thisRocket[0].description } </div>
        <ImagesArea
          title={thisRocket[0].rocket_name}
          images={thisRocket[0].flickr_images}
        />
      </Section>
    );
  }
}

RocketView.propTypes = {
  rocketCollection: PropTypes.shape({
    rockets: PropTypes.arrayOf(PropTypes.shape({
      rocket_id: PropTypes.string,
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getRockets: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      rocket_id: PropTypes.string,
    }),
  }).isRequired,
}

const mapStateToProps = state => ({
  rocketCollection: state.rocketCollection,
});

const mapDispatchToProps = {
  getRockets: getRocketsThunk,
};

const ConnectedRocketView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RocketView);

export default ConnectedRocketView;
