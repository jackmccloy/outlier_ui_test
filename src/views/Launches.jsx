import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getLaunchesThunk } from "../actions/Launches";
import Launch from '../components/Launch';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const LaunchesH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class LaunchesView extends Component {
  componentDidMount() {
    const { getLaunches } = this.props;
    getLaunches();
  }

  launchesArea() {
    const { launchCollection } = this.props;

    if (!launchCollection || launchCollection.fetching) {
      return <div> LOADING... </div>;
    }

    if (launchCollection.error) {
      return <div> { launchCollection.error } </div>
    }

    if (!launchCollection.launches.length) {
      return <div> NO DATA </div>;
    }

    const launches = launchCollection.launches.map(launch => (
      <Launch
        key={launch.flight_number}
        launch={launch}
      />
    ));

    return <ul>{launches}</ul>;
  }

  render() {
    return (
      <Section>
        <LaunchesH1> SpaceX launches </LaunchesH1>
        {this.launchesArea()}
      </Section>
    );
  }
}

LaunchesView.propTypes = {
  launchCollection: PropTypes.shape({
    launches: PropTypes.arrayOf(PropTypes.shape({
      flight_number: PropTypes.number,
      mission_name: PropTypes.string,
      rocket: PropTypes.shape({
        rocket_id: PropTypes.string,
        rocket_name: PropTypes.string,
      }),
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getLaunches: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  launchCollection: state.launchCollection,
});

const mapDispatchToProps = {
  getLaunches: getLaunchesThunk,
};

const ConnectedLaunchesView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(LaunchesView);

export default ConnectedLaunchesView;
