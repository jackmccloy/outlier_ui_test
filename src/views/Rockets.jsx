import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { connect } from "react-redux";

import styled from 'styled-components';

import { getRocketsThunk } from "../actions/rockets";
import Rocket from '../components/Rocket';

const Section = styled.section`
  flex: 1;
  overflow: auto;
  padding: 20px;
`;

const RocketsH1 = styled.h1`
  margin-top: 0;
  border-bottom: 2px solid #999;
  color: #666;
`

class RocketsView extends Component {
  componentDidMount() {
    const { getRockets } = this.props;
    getRockets();
  }

  rocketsArea() {
    const { rocketCollection } = this.props;

    if (!rocketCollection || rocketCollection.fetching) {
      return (
        <div> LOADING... </div>
      );
    }

    if (rocketCollection.error) {
      return (
        <div> { rocketCollection.error } </div>
      );
    }

    if (!rocketCollection.rockets.length) {
      return (
        <div> NO DATA </div>
      );
    }

    const rockets = rocketCollection.rockets.map(rocket => (
      <Rocket
        key={rocket.rocket_id}
        rocket={rocket}
      />
    ));

    return (
      <ul>{rockets}</ul>
    );
  }

  render() {
    return (
      <Section>
        <RocketsH1> SpaceX rockets </RocketsH1>
        {this.rocketsArea()}
      </Section>
    );
  }
}

RocketsView.propTypes = {
  rocketCollection: PropTypes.shape({
    rockets: PropTypes.arrayOf(PropTypes.shape({
      rocket_id: PropTypes.string,
    })),
    fetching: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  getRockets: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  rocketCollection: state.rocketCollection,
});

const mapDispatchToProps = {
  getRockets: getRocketsThunk,
};

const ConnectedRocketsView = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RocketsView);

export default ConnectedRocketsView;
