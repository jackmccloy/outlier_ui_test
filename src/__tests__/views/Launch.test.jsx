import React from 'react';

import { mount } from 'enzyme';

import renderer from 'react-test-renderer';

import {
  MemoryRouter,
} from 'react-router';

import { LaunchView } from '../../views/Launch';

import { launchApiResponse } from '../../__mocks__/apiResponses';

describe('<LaunchesView />', () => {
  let getLaunches;

  beforeEach(() => {
    getLaunches = jest.fn();

    mount(
      <LaunchView
        getLaunches={getLaunches}
        launchCollection={{
          launches: [],
          fetching: false,
          error: '',
        }}
        match={{
          params: {
            flight_number: '1',
          }
        }}
      />,
    );
  });

  it('calls getLaunches on mount', () => {
    expect(getLaunches.mock.calls.length).toBe(1);
  });

  it('renders correctly while fetching', () => {
    const tree = renderer
      .create(
        <LaunchView
          getLaunches={jest.fn()}
          launchCollection={{
            launches: [],
            fetching: true,
            error: '',
          }}
          match={{
            params: {
              flight_number: '1',
            }
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly after fetch error', () => {
    const tree = renderer
      .create(
        <LaunchView
          getLaunches={jest.fn()}
          launchCollection={{
            launches: [],
            fetching: false,
            error: 'An error occured',
          }}
          match={{
            params: {
              flight_number: '1',
            }
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with data', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <LaunchView
            getLaunches={jest.fn()}
            launchCollection={{
              launches: launchApiResponse,
              fetching: false,
              error: '',
            }}
            match={{
              params: {
                flight_number: '1',
              }
            }}
          />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
