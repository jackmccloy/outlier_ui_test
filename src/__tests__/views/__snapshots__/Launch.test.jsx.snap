// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchesView /> renders correctly after fetch error 1`] = `
<section
  className="Launch__Section-sc-10cqa0r-0 iEYcVm"
>
  <h1
    className="Launch__LaunchH1-sc-10cqa0r-1 fkUYmW"
  >
     SpaceX 
  </h1>
  <div>
     
    An error occured
     
  </div>
</section>
`;

exports[`<LaunchesView /> renders correctly while fetching 1`] = `
<section
  className="Launch__Section-sc-10cqa0r-0 iEYcVm"
>
  <h1
    className="Launch__LaunchH1-sc-10cqa0r-1 fkUYmW"
  >
     SpaceX 
  </h1>
  <div>
     LOADING... 
  </div>
</section>
`;

exports[`<LaunchesView /> renders correctly with data 1`] = `
<section
  className="Launch__Section-sc-10cqa0r-0 iEYcVm"
>
  <h1
    className="Launch__LaunchH1-sc-10cqa0r-1 fkUYmW"
  >
     SpaceX 
    FalconSat
    , 
    2006
  </h1>
  <h2>
     Rocket: 
    <a
      href="/rockets/falcon1"
      onClick={[Function]}
    >
      Falcon 1
    </a>
  </h2>
  <div>
     
    Engine failure at 33 seconds and loss of vehicle
     
  </div>
</section>
`;
